core/logger/audit/audit_logger.go

```diff
@@ -0,0 +1,231 @@
+	logger          logger.Logger
+	enabled         bool
+	serviceURL      string
+	serviceHeaders  []serviceHeader
+	jsonWrapperKey  string
+	environmentName string
+	hostname        string
+	localIP         string
+	loggingChannel  chan WrappedAuditLog
+}
+
+// Configurable headers to include in POST to log service
+type serviceHeader struct {
+	header string
+	value  string
+}
+
+type WrappedAuditLog struct {
+	eventID EventID
+	data    Data
+}
+
+// NewAuditLogger returns a buffer push system that ingests audit log events and
+// asynchronously pushes them up to an HTTP log service.
+// Parses and validates the AUDIT_LOGS_* environment values and returns an enabled
+// AuditLogger instance. If the environment variables are not set, the logger
+// is disabled and short circuits execution via enabled flag.
+func NewAuditLogger(logger logger.Logger, config *AuditLoggerConfig) (AuditLogger, error) {
+	if config == nil {
+		logger.Info("Audit logger configuration is nil. Cannot start audit logger subsystem and audit events will not be captured.")
+		return &AuditLoggerService{}, nil
+	}
+
+	hostname, err := os.Hostname()
+	if err != nil {
+		return &AuditLoggerService{}, errors.Errorf("Audit Log initialization error - unable to get hostname: %s", err)
+	}
+
+	loggingChannel := make(chan WrappedAuditLog, bufferCapacity)
+
+	// Finally, create new auditLogger with parameters
+	auditLogger := AuditLoggerService{
+		logger:          logger.Helper(1),
+		enabled:         true,
+		serviceURL:      config.forwardToUrl,
+		serviceHeaders:  config.headers,
+		jsonWrapperKey:  config.jsonWrapperKey,
+		environmentName: config.environment,
+		hostname:        hostname,
+		localIP:         getLocalIP(),
+		loggingChannel:  loggingChannel,
+	}
+
+	// Start our go routine that will receive logs and send them out to the
+	// configured service.
+	go auditLogger.auditLoggerRoutine()
+
+	return &auditLogger, nil
+}
+
+// / Entrypoint for new audit logs. This buffers all logs that come in they will
+// / sent out by the goroutine that was started when the AuditLoggerService was
+// / created. If this service was not enabled, this immeidately returns.
+// /
+// / This function never blocks.
+func (l *AuditLoggerService) Audit(ctx context.Context, eventID EventID, data Data) {
+	if !l.enabled {
+		return
+	}
+
+	wrappedLog := WrappedAuditLog{
+		eventID: eventID,
+		data:    data,
+	}
+
+	select {
+	case l.loggingChannel <- wrappedLog:
+	default:
+		if l.loggingChannel == nil {
+			l.logger.Errorw("Could not send log to audit subsystem because it has gone away!")
+		} else {
+			l.logger.Errorw("Audit log buffer is full. Dropping log with eventID: %s", eventID)
+		}
+	}
+}
+
+// / Entrypoint for our log handling goroutine. This waits on the channel and sends out
+// / logs as they come in.
+// /
+// / This function calls postLogToLogService which blocks.
+func (l *AuditLoggerService) auditLoggerRoutine() {
+	for event := range l.loggingChannel {
```

Comment:
Please implement as a `service.ServiceCtx` that responds to `Close`.

For a (relatively) simple example, see `EthResender`.